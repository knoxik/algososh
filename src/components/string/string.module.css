.wrapper {
    display: flex;
    flex-direction: row;
    column-gap: 12px;
    justify-content: center;
}

.input {
    max-width: 377px;
}

.circles {
    margin-top: 120px;
    display: flex;
    flex-direction: row;
    column-gap: 16px;
    justify-content: center;
}