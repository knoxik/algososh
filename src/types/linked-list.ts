export class Node<T> {
    value: T
    next: Node<T> | null
    constructor(value: T, next?: Node<T> | null) {
      this.value = value;
      this.next = (next === undefined ? null : next);
    }
}

interface ILinkedList<T> {
    append: (element: T) => void;
    prepend: (element: T) => void;
    insertAt: (element: T, position: number) => void;
    getSize: () => number;
}

class LinkedList<T> implements ILinkedList<T> {
    private head: Node<T> | null;
    private tail: Node<T> | null;
    private size: number;
    constructor() {
      this.head = null;
      this.tail = null;
      this.size = 0;
    }

    clear() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    getHead() {
        return this.head
    }

    getTail() {
        return this.tail
    }
  
    insertAt(element: T, index: number) {
      if (index < 0 || index > this.size) {
        return false;
      } else {
        const node = new Node(element);
  
        if (index === 0) {
          this.prepend(element);
          return true;
        } else {
          this.size++;
          let curr = this.head;
          let currIndex = 0;
          
          while(curr) {
            if (currIndex === index - 1) {
              if (curr.next !== null) {
                node.next = curr.next
                curr.next = node;
                return true;
              } else {
                curr.next = node;
                return true;
              }
            } else {
              curr = curr.next;
              currIndex++;
              
            }
          }
        }
      }
    }

    deleteAt(index: number) {
        if (index < 0 || index > this.size - 1) {
            return false;
        }
        
        if (this.size === 2) {
            if (index === 0) {
              return this.deleteHead()
            }
            if (index > 0) {
              return this.deleteTail()
            }
        }

        let removalType
        if (index === 0) {
            removalType = 'head'
        } else if (index >= this.size) {
            removalType = 'tail'
        } else {
            removalType = 'middle'
        }

        if (removalType === 'head') {
            return this.deleteHead()
        }

        if (removalType === 'tail') {
            return this.deleteTail()
        }

        if (removalType === 'middle') {
            const preIdx = this.traverseToIndex(index - 1)
            if (preIdx) {
                const targetIdx = preIdx.next
                if (targetIdx) {
                    preIdx.next = targetIdx.next
                    this.size--
                } 
            }
            
            return true;
        }
    }

    traverseToIndex(index: number) {
        let counter = 0;
        let currentNode = this.head;
        
        while (counter !== index) {
          if (currentNode) {
            currentNode = currentNode.next
            counter++
          }
        }
    
        return currentNode
      }
  
    append(element: T) {
      const node = new Node(element);
  
      if (this.head === null) {
        this.head = node;
        this.tail = node;
      } else {
        if (this.tail) {
            this.tail.next = node;
            this.tail = node;
        }       
      }
      this.size++;
    }

    prepend(element: T) {
        const node = new Node(element);
        node.next = this.head;
        this.head = node;
        this.size++;
    }

    deleteTail() {
        if (!this.tail || !this.head) return false;

        if (this.head === this.tail) {
            this.head = null
            this.tail = null
            this.size--
            return true
        }

        let currentNode = this.head
        while (currentNode.next) {
            if (!currentNode.next.next) {
                currentNode.next = null
            } else {
                currentNode = currentNode.next
            }
        }

        this.tail = currentNode
        this.size--
        return true;
    }

    deleteHead() {
        if (!this.head || !this.tail) return false;

        if (this.head === this.tail) {
            this.head = null
            this.tail = null
            this.size--
            return true;
        }

        const newHead = this.head.next
        this.head = newHead
        this.size--
        return true
    }

  
    getSize() {
      return this.size;
    }
}

export const linkedList = new LinkedList<string>();