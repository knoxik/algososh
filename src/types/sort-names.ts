export enum SortNames {
    Selection = "selection",
    Bubble = "bubble",
  }
  