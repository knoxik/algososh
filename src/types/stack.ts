export interface IStack<T> {
    push: (item: T) => void;
    pop: () => void;
    peak: () => T | null;
    getSize: () => number;
}

export class Stack<T> implements IStack<T> {
    private container: T[] = [];
  
    push = (item: T): void => {
      this.container.push(item)
    };
  
    pop = (): void => {
      this.container.pop()
    };
  
    peak = (): T | null => {
      const len = this.getSize();
      if (len === 0) {
        return null;
      }
      return this.container[len - 1];
    };

    getAll = (): T[] => {
        return this.container;
    }

    clear = (): void => {
        this.container = [];
    }
  
    getSize = () => this.container.length;
  }

export const stack = new Stack<string>()