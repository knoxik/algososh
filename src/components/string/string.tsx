import React, { ReactElement } from "react";
import { SolutionLayout } from "../ui/solution-layout/solution-layout";
import { Input } from "../ui/input/input";
import { Button } from "../ui/button/button";
import { Circle } from "../ui/circle/circle";
import Styles from './string.module.css'
import { ChangeEvent, MouseEvent } from 'react';
import { ElementStates } from "../../types/element-states";
import { DELAY_IN_MS } from "../../constants/delays";
import { CircleProps } from "../ui/circle/circle";


export const StringComponent: React.FC = () => {
  const [disabled, setDisabled] = React.useState(true);
  const [loader, setLoader] = React.useState(false);
  const [visible, setVisible] = React.useState(false);
  const [text, setText] = React.useState('');
  const [circleOptions, setCircleOptions] = React.useState<CircleProps[]>([{}])
  const { v4: uuidv4 } = require('uuid');

  const onChange = (evt: ChangeEvent<HTMLInputElement>) => {
    const target = evt.target;
    setDisabled(target.value.length <= 0);
    setText(target.value);
  }

  const onClick = (evt: MouseEvent<HTMLButtonElement>) => {
    let str = text;
    setLoader(true);
    setText('');

    let circleOptions: CircleProps[] = [];

    str.split('').forEach((char, index) => {
      circleOptions.push({
        letter: char.toUpperCase(),
        state: index===str.length-1 || index===0  ? ElementStates.Changing : ElementStates.Default,
      })
    })
    setCircleOptions([...circleOptions]);
    setVisible(true);

    reverseArrByStep([...circleOptions], 0, circleOptions.length - 1);
  }

  const reverseArrByStep = (arr: CircleProps[], head: number, tail: number) => {
    if (tail > head) {
      setTimeout(function(){
        let temp = arr[head];
        arr[head] = arr[tail];
        arr[tail] = temp;

        arr[head].state = ElementStates.Modified
        arr[tail].state = ElementStates.Modified

        head++;
        tail--;

        arr[head].state = ElementStates.Changing
        arr[tail].state = ElementStates.Changing

        setCircleOptions([...arr]);
        reverseArrByStep(arr, head, tail);
      }, DELAY_IN_MS);

    } else {
        if (tail <= head) {
          arr[head].state = ElementStates.Modified
          arr[tail].state = ElementStates.Modified
          setCircleOptions([...arr]);
        }
        setDisabled(true);
        setLoader(false);
        
    } 
  }

  return (
    <SolutionLayout title="Строка">
      <div className={Styles.wrapper}>
        <Input value={text} isLimitText={true} maxLength={11} extraClass={Styles.input} onChange={onChange}/>
        <Button text='Развернуть' disabled={disabled} isLoader={loader} onClick={onClick} />
      </div>

      {visible && (
        <div className={Styles.circles}>
          {
            circleOptions.map((option) => (
              <Circle key={uuidv4()} letter={option.letter} state={option.state}/>
            ))
          }
        </div>
      )}
      
    </SolutionLayout>
  );
};
