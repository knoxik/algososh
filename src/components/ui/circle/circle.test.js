import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';

import { Circle } from './circle';
import { ElementStates } from '../../../types/element-states';
import { StringComponent } from '../../string/string';

describe('Проверка circle', () => {
    it('Circle без буквы', () => {
        const CircleComponent = renderer.create(<Circle />).toJSON();
        expect(CircleComponent).toMatchSnapshot();
    }) 

    it('Circle с буквой', () => {
        const CircleComponent = renderer.create(<Circle letter='a' />).toJSON();
        expect(CircleComponent).toMatchSnapshot();
    }) 

    it('Circle с head', () => {
        const CircleComponent = renderer.create(<Circle head='1'/>).toJSON();
        expect(CircleComponent).toMatchSnapshot();
    }) 

    it('Circle с react-элементом в head', () => {
        const CircleComponent = renderer.create(<Circle head={<Circle/>}/>).toJSON();
        expect(CircleComponent).toMatchSnapshot();
    }) 

    it('Circle с tail', () => {
        const CircleComponent = renderer.create(<Circle tail='1'/>).toJSON();
        expect(CircleComponent).toMatchSnapshot();
    }) 

    it('Circle с react-элементом в tail', () => {
        const CircleComponent = renderer.create(<Circle tail={<Circle />}/>).toJSON();
        expect(CircleComponent).toMatchSnapshot();
    }) 

    it('Circle с index', () => {
        const CircleComponent = renderer.create(<Circle index={1}/>).toJSON();
        expect(CircleComponent).toMatchSnapshot();
    }) 

    it('Circle с пропом isSmall ===  true', () => {
        const CircleComponent = renderer.create(<Circle isSmall={true}/>).toJSON();
        expect(CircleComponent).toMatchSnapshot();
    }) 

    it('Circle в состоянии default', () => {
        const CircleComponent = renderer.create(<Circle state={ElementStates.Default}/>).toJSON();
        expect(CircleComponent).toMatchSnapshot();
    }) 

    it('Circle в состоянии changing', () => {
        const CircleComponent = renderer.create(<Circle state={ElementStates.Changing}/>).toJSON();
        expect(CircleComponent).toMatchSnapshot();
    }) 

    it('Circle в состоянии modified', () => {
        const CircleComponent = renderer.create(<Circle state={ElementStates.Modified}/>).toJSON();
        expect(CircleComponent).toMatchSnapshot();
    })     
});