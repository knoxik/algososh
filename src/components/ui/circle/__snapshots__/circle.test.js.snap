// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Проверка circle Circle без буквы 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
      data-testid="circle-letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка circle Circle в состоянии changing 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   changing"
  >
    <p
      className="text text_type_circle text_color_input letter"
      data-testid="circle-letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка circle Circle в состоянии default 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
      data-testid="circle-letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка circle Circle в состоянии modified 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   modified"
  >
    <p
      className="text text_type_circle text_color_input letter"
      data-testid="circle-letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка circle Circle с head 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head string"
  >
    1
  </div>
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
      data-testid="circle-letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка circle Circle с index 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
      data-testid="circle-letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  >
    1
  </p>
  <div
    className="text text_type_input text_color_input mt-4 absolute tail60 element"
  />
</div>
`;

exports[`Проверка circle Circle с react-элементом в head 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  >
    <div
      className="content "
    >
      <div
        className="text text_type_input text_color_input mb-4 absolute head element"
      />
      <div
        className="circle   default"
      >
        <p
          className="text text_type_circle text_color_input letter"
          data-testid="circle-letter"
        />
      </div>
      <p
        className="text text_type_input text_color_input mt-4 absolute index"
      />
      <div
        className="text text_type_input text_color_input mt-4 absolute tail30 element"
      />
    </div>
  </div>
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
      data-testid="circle-letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка circle Circle с react-элементом в tail 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
      data-testid="circle-letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  >
    <div
      className="content "
    >
      <div
        className="text text_type_input text_color_input mb-4 absolute head element"
      />
      <div
        className="circle   default"
      >
        <p
          className="text text_type_circle text_color_input letter"
          data-testid="circle-letter"
        />
      </div>
      <p
        className="text text_type_input text_color_input mt-4 absolute index"
      />
      <div
        className="text text_type_input text_color_input mt-4 absolute tail30 element"
      />
    </div>
  </div>
</div>
`;

exports[`Проверка circle Circle с tail 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
      data-testid="circle-letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 string"
  >
    1
  </div>
</div>
`;

exports[`Проверка circle Circle с буквой 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
      data-testid="circle-letter"
    >
      a
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка circle Circle с пропом isSmall ===  true 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle  small default"
  >
    <p
      className="text text_type_circle text_color_input letter"
      data-testid="circle-letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;
